platform/lang-impl/src/com/intellij/openapi/roots/impl/PushedFilePropertiesUpdaterImpl.java

```diff
@@ -303,46 +304,50 @@ public boolean processFile(final VirtualFile fileOrDir) {
       tasks.add(iteration);
     }
 
-    invoke2xConcurrentlyIfPossible(tasks);
+    invokeConcurrentlyIfPossible(tasks);
   }
 
-  public static void invoke2xConcurrentlyIfPossible(final List<Runnable> tasks) {
+  public static void invokeConcurrentlyIfPossible(final List<Runnable> tasks) {
     if (tasks.size() == 1 ||
         ApplicationManager.getApplication().isWriteAccessAllowed() ||
         !Registry.is("idea.concurrent.scanning.files.to.index")) {
       for(Runnable r:tasks) r.run();
       return;
     }
-    
+
     final ProgressIndicator progress = ProgressManager.getInstance().getProgressIndicator();
     
     final ConcurrentLinkedQueue<Runnable> tasksQueue = new ConcurrentLinkedQueue<Runnable>(tasks);
-    Future<?> result = null;
-    if (tasks.size() > 1) {
-      result = ApplicationManager.getApplication().executeOnPooledThread(new Runnable() {
-        @Override
-        public void run() {
-          ProgressManager.getInstance().runProcess(new Runnable() {
-            @Override
-            public void run() {
-              Runnable runnable;
-              while ((runnable = tasksQueue.poll()) != null) runnable.run();
-            }
-          }, ProgressWrapper.wrap(progress));
-        }
-      });
+    List<Future<?>> results = Lists.newArrayList();
+    int numThreads = Math.min(CacheUpdateRunner.indexingThreadCount() - 1, tasks.size() - 1);
```

Comment:
Some comments (for consideration):
- for 2 cpu system (my MacBook Air or our testing servers) there is no concurrency right now
- if we open several projects at once (e.g. 3) we'll create too many workers (it still can happen with previous 2x concurrent code).
- I do not observe significant speed up on my Windows machine with the change. There is one more thing to consider for this place: if we open completely new project (not existing one) the code will do real indexing of files and such indexing is 2x concurrent as well. Adding extra threads for this phase will not increase speed right now at all.  

So I'd suggest to add BoundedExecutor (over PooledThreads in Application) that will remove problems with several projects.